[AIRFLOW-1076] Add get method for template variable accessor

dhuang opened the pull request:
Support getting variables in templates by string. This is necessary when
fetching variables with characters not allowed in a class attribute
name. We can then also support returning default values when a variable does
not exist.

Original PR went stale, https://github.com/apache/airflow/pull/2223.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-1076

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
  - See above.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
  - Added unit tests for calling `var.value.get()` and `var.json.get()`, with or without default

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.

dimberman reviewed (approved):
This LGTM, but I want a second approval since this is a front-facing change. @ashb can you give this a look? Also I imagine this would be an airflow 2.0 change.

ashb commented on docs/macros-ref.rst:
```diff
@@ -88,7 +88,12 @@ attributes and methods.
+``{{ var.json.my_dict_var.key1 }}``.
+
+It is also possible to fetch a variable by string if needed with
+``{{ var.value.get('my.var', 'fallback') }}`` or
```
Instead of needing a code change can we instead use Jinja's built in `default` fliter? `{{ var.value['my.var'] | default 'fallback' }}`?

https://jinja.palletsprojects.com/en/2.10.x/templates/#default

dhuang replied:
This resolves my original reason for adding this method (periods in variable keys), but that default logic doesn't work because it throws a KeyError first when it's missing. `Error rendering template: 'Variable my.var does not exist'` is the error I get from your example.

ashb replied:
Ah okay, yeah I thought Jinja might catpure a KeyError and handle it for the default filter, but I guess not. Cool, just wanted to check if this PR was needed, and it looks like it is. :+1:

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1178,34 +1178,60 @@ def get_template_context(self, session=None) -> Dict[str, Any]:
+            def __getattr__(
+                self,
+                item: str,
+                default_var: Any = Variable._Variable__NO_DEFAULT_SENTINEL,
```
It is impossible to call this with anything other than this default value except by doing `var.__getattr__(name, default_value)` so we should just remove these changes to `__getattr__`

dhuang replied:
yep had thought I was actually using `_getattr_` when I was doing `self.var` in `get()`, removed.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1178,34 +1178,60 @@ def get_template_context(self, session=None) -> Dict[str, Any]:
+                item: str,
+                default_var: Any = Variable._Variable__NO_DEFAULT_SENTINEL,
+            ):
+                self.var = Variable.get(item, default_var=default_var, deserialize_json=True)
```
This is a static method, there is no `self` in local scope, so `self` is actually the TaskInstance. This is not what you want.

```suggestion
                return Variable.get(item, default_var=default_var, deserialize_json=True)
```

dhuang replied:
🤦‍♂ blindly copy pasted from `__getattr_`, fixed up.

ashb commented on tests/test_core.py:
```diff
@@ -588,6 +588,42 @@ def verify_templated_field(context):
+            some_templated_field='{{ var.value.get("missing_variable", "fallback") }}',
+            dag=self.dag)
+        t.execute = verify_templated_field
+        t.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE, ignore_ti_state=True)
```
Running the operator is overkill (yes I see you copied this existing pattern). Instead of putting these tests here look at the tests in tests/models/test_baseoperator.py which calls `task.render_template` more directly.

dhuang replied:
Yep that looks much cleaner, plus parametrized them as well. Also moved to `tests/models/test_taskinstance.py`, but let me know if it belongs elsewhere or we want to keep it in core.

ashb reviewed (changes requested):
Have left comments, but this may not need any code change at all. Could you check and see if we actuall need this code change at all?

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1178,34 +1178,60 @@ def get_template_context(self, session=None) -> Dict[str, Any]:
-            {var.value.your_variable_name}.
+            Wrapper around Variable. This way you can get variables in
+            templates by using {{ var.value.variable_name }} or
+            {{ var.value.get('variable_name', 'fallback') }}.
```
```suggestion
            ``{{ var.value.get('variable_name', 'fallback') }}``.
```

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1178,34 +1178,60 @@ def get_template_context(self, session=None) -> Dict[str, Any]:
-            Wrapper around Variable. This way you can get variables in templates by using
-            {var.value.your_variable_name}.
+            Wrapper around Variable. This way you can get variables in
+            templates by using {{ var.value.variable_name }} or
```
```suggestion
            templates by using ``{{ var.value.variable_name }}`` or
```

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1178,34 +1178,60 @@ def get_template_context(self, session=None) -> Dict[str, Any]:
-            Wrapper around deserialized Variables. This way you can get variables
-            in templates by using {var.json.your_variable_name}.
+            Wrapper around Variable. This way you can get variables in
+            templates by using {{ var.json.variable_name }} or
```
```suggestion
            templates by using ``{{ var.json.variable_name }}`` or
```

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1178,34 +1178,60 @@ def get_template_context(self, session=None) -> Dict[str, Any]:
-            in templates by using {var.json.your_variable_name}.
+            Wrapper around Variable. This way you can get variables in
+            templates by using {{ var.json.variable_name }} or
+            {{ var.json.get('variable_name', {'fall': 'back'}) }}.
```
```suggestion
            ``{{ var.json.get('variable_name', {'fall': 'back'}) }}``.
```

dhuang committed:
[AIRFLOW-1076] Add get method for template variable accessor

Support getting variables in templates by string. This is necessary when
fetching variables with characters not allowed in a class attribute
name. We can then also support returning default values when a variable does
not exist.

dhuang reviewed (commented):
Tried your suggestion to avoid code change. I don't think it's possible, but let me know if you have other ideas! Otherwise, comments addressed.

ashb commented on tests/test_core.py:
```diff
@@ -568,67 +568,6 @@ def verify_templated_field(context):
-            some_templated_field='{{ var.value.a_variable }}',
-            dag=self.dag)
-        t.execute = verify_templated_field
-        t.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE, ignore_ti_state=True)
```
Yup, nice one! 💯
